ci/submodule-sync.sh

```diff
@@ -0,0 +1,82 @@
+#!/bin/bash
+#
+# Copyright (c) 2022, NVIDIA CORPORATION. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# NOTE:
+#     this script is for jenkins only, and should not be used for local development
+#     run with ci/Dockerfile in jenkins:
+#         scl enable devtoolset-9 rh-python38 "ci/submodule-sync.sh"
+
+set -ex
+
+OWNER=${OWNER:-"NVIDIA"}
+REPO=${REPO:-"spark-rapids-jni"}
+PARALLEL_LEVEL=${PARALLEL_LEVEL:-4}
+REPO_LOC="github.com/${OWNER}/${REPO}.git"
+
+GIT_AUTHOR_NAME="spark-rapids automation"
+GIT_COMMITTER_NAME="spark-rapids automation"
+GIT_AUTHOR_EMAIL="70000568+nvauto@users.noreply.github.com"
+GIT_COMMITTER_EMAIL="70000568+nvauto@users.noreply.github.com"
+git submodule update --init --recursive
+
+INTERMEDIATE_HEAD=bot-submodule-sync-${REF}
+cudf_prev_sha=$(git -C thirdparty/cudf rev-parse HEAD)
+git checkout -b ${INTERMEDIATE_HEAD} origin/${REF}
+# sync up cudf from remote
+git submodule update --remote --merge
+cudf_sha=$(git -C thirdparty/cudf rev-parse HEAD)
+if [[ "${cudf_sha}" == "${cudf_prev_sha}" ]]; then
+  echo "Submodule is up to date."
+  exit 0
+fi
+
+echo "Try update cudf submodule to ${cudf_sha}..."
+git add .
+git commit -s -m "Update submodule cudf to ${cudf_sha}"
+sha=$(git rev-parse HEAD)
+
+echo "Test against ${cudf_sha}..."
+
+set +e
+mvn verify \
+  -DCPP_PARALLEL_LEVEL=${PARALLEL_LEVEL} \
+  -Dlibcudf.build.configure=true \
+  -DUSE_GDS=ON
+ret="$?"
+set -e
+
+test_pass="False"
+if [[ "${ret}" == "0" ]]; then
+  echo "Test failed, will update the result"
+  test_pass="True"
+else
+  echo "Test passed, will try merge the change"
+fi
+
+# force push the intermediate branch and create PR against REF
```

Comment:
I'm still a little unclear why we need to force push instead of just push.  Is it because we're sharing the same branch across multiple sync PRs?  Even if we share the branch, I'm not sure why subsequent pushes wouldn't always be essentially a fast-forward of the previous version on that branch.

Couple of questions:
- If there is the potential of sharing, should we ensure each upmerge attempt uses a separate branch to avoid the potential of accidentally colliding?  Seems like we would never need to force-push in that scenario.
- How are these intermediate branches cleaned up?

Reply:
> If there is the potential of sharing, should we ensure each upmerge attempt uses a separate branch to avoid the potential of accidentally colliding? Seems like we would never need to force-push in that scenario.

For current setup, there should be exactly one branch (bot-auto-merge-<BASE>) for BASE branch's submodule sync which also means there should be zero or only one opened PR for BASE branch's submodule sync, all updates will go into the same branch to avoid opening multiple sync PRs. And also the single sync branch will be manually removed after we disable the corresponding sync job
 
The reason I decide to force push here is
1. If someone (not sync job) accidentally modified the sync branch, we may resolve the potential conflict
2. the sync job would need to do an extra upmerge for current sync branch instead of always simply fetching from remote each time, then try test and update submodule commit. And we may need to use the `squash merge` for this PR, so auto-merge conflict fix for this kind of PR may require another steps (conflict fix for submodule sync should not happen frequently, though)
3. The comment history should help leave the test results in thread, we do not have to keep the commit history here

Yes, to use the force push is not a must but for simplifying the workflow. I can make it to just push, and some potential changes could be,
1. every job fetch both remote **sync branch** and BASE branch
2. merge BASE into sync branch to make sure local test is up to date
3. use `squash` merge strategy for the PR to avoid merge independent failed-test cudf commit

@jlowe Do you think this one looks better? Thanks~

Reply:
> If someone (not sync job) accidentally modified the sync branch, we may resolve the potential conflict

This is one of the main reasons I'm _not_ a fan of the force-push approach.  The force-push can silently destroy changes on the branch.  I would prefer if this created a new, temporary branch for the PR, then there's no need to force-push (or if the branch already exists we _want it to fail_ to avoid clobbering any other changes).  When the PR merges, the action merging would then also clean up the temporary branch.  We would need a way to resolve cleaning up the temporary branches when the PR is closed in other ways (e.g.: by resolving the merge conflict manually, etc.).  Maybe always delete the temporary branch via a github trigger when the corresponding auto-update PR is closed (even if not merged)?

> And we may need to use the squash merge for this PR

As for squashing instead of merging the PR, that seems appropriate here since this should be treated just like a regular development PR which we would also squash.

Reply:
> Maybe always delete the temporary branch via a github trigger when the corresponding auto-update PR is closed (even if not merged)?

Yeah, the github could add the strategy to automatically delete the HEAD branch after PR gets merged, but it cannot tell which one is from auto bot. Let me try simply add some code to clean up the temp branch if `auto-merged` succeed.

I will also refactor the process here to not force push